Fix various problems with types in expression context

slavapestov opened the pull request:
The parser doesn't know if a dot expression references a member or a type. This creates an ambiguity with array literals and some other cases. We resolve these ambiguities with a pre-check pass which attempts to fold certain expressions to types.

This now works better. Examples that did not work before:

```
_ = [A.B]()
_ = [Outer<Int>.Inner<String>]()
_ = [SomeClass<Int> & MyProto]()
```

Also, patch up a few cases in Swift 3 mode where '.self' was previously not required to emit a warning, for example:

```
_ = Swift.Int
foo(Swift.Int, 0)
```

In Swift 4 mode, omission of '.self' is now an error in these cases. Unfortunately there are still some holes in the '.self' check.

There is a fair bit of refactoring here; see the individual commit messages for details. Basically the code paths for type lookup in expression context and type lookup in type context each had their own quirks; I attempted to reconcile the differences and unify the implementations more while maintaining compatibility with old behaviors.

Fixes <rdar://problem/16849958> and <https://bugs.swift.org/browse/SR-349>.

slavapestov committed:
Sema: Support member types in preCheckExpression()

This folds member access on types to a TypeExpr if the member
resolves to a nested type.

This allows [Foo.Bar]() to become an ApplyExpr of a TypeExpr,
rather than an ApplyExpr of an array literal; previously,
only [Foo]() worked.

Two cases that are still unsupported:

1) If G is a generic type and T is a generic typealias, then
   G.T<X> cannot be folded to a TypeExpr, because we cannot
   represent a generic typealias with an unbound generic
   parent type. Such lookups remain member lookups, where
   Sema first opens the base type to produce a bound generic
   type G<$T0>, and then resolves the generic typealias
   member on that.

2) If T is a generic parameter and X is an associated type,
   T.X is not folded down to a TypeExpr either.

Fixes <rdar://problem/16849958>.

slavapestov committed:
Sema: Specialization of nested types in expression context

Teach preCheckExpression() about UnresolvedSpecializeExpr
where the base is a TypeExpr.

This allows us to type check expressions like
'[Outer<T>.Inner<U>]()' by folding them down to a TypeExpr
with an array type.

slavapestov commented:
@DougGregor This isn't quite ready to go in because it appears that a few places where we didn't require '.self' now require it. I'll have to cook up a compatibility hack :)

DougGregor commented on lib/Sema/TypeCheckConstraints.cpp:
```diff
@@ -1048,6 +1051,127 @@ bool PreCheckExpression::walkToClosureExprPre(ClosureExpr *closure) {
+              BaseTy->isExistentialType())
+            return nullptr;
+
+          // 2) Generic typealias reference with unbound generic base.
```
It seems dicey to filter these out at this level, but I assume you're dodging bugs elsewhere?

slavapestov replied:
I'm going to refactor things so that the "filtering" isn't necessary. You're right that it's fragile.

slavapestov committed:
Sema: Clean up openGeneric() for nested generic types

If we had an unbound generic type whose parent was a bound
generic type, and the outer type's generic signature
placed generic constraints on outer parameters, we would
create type variables that weren't ever bound to anything.

Previously, this would never come up, but it will once
preCheckExpression() is folding more MemberRefExprs down
to TypeExprs.

slavapestov committed:
Sema: Fix for typealiases in protocols not getting a generic environment

A typealias in a protocol may have had an interface type but no
generic signature. This made ConstraintSystem::openGeneric()
fail with a typealias.

A later patch makes us call openGeneric() in more cases, and some
crashers regress without this fix.

slavapestov committed:
Sema: More consistent enforcement of '.self' on metatype values in Swift 4 mode

In Swift 3, we required that '.self' be specified on TypeExprs
and DeclRefExprs that reference types.

However, types referenced as member lookups, such as 'Foo.Bar',
did not get this treatment, and '.self' was not required.

Fix this by emitting warnings in the cases that Swift 3 did not
diagnose, and producing errors in Swift 4 mode where we want
strict enforcement.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov committed:
Sema: Move some code from preCheckExpression() pass to static methods on TypeExpr

This is just an NFC refactoring to simplify the pre-check
pass a little bit.

slavapestov committed:
Sema: Factor out some duplication in member type lookup

When resolving a nested identifier, we would take different
code paths depending on whether the identifier already had
a declaration bound or not. These two code paths had to
diagnose the same set of problems, so unify them.

This also lets us rip out similar checks in preCheckExpression(),
which can now rely on the diagnostics being produced consistently
when constructing TypeReprs with already-bound decls.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test
